homeassistant/components/xiaomi_miio/__init__.py

```diff
@@ -119,6 +129,24 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up the Xiaomi Miio components from a config entry."""
     hass.data.setdefault(DOMAIN, {})
+    host = entry.data[CONF_HOST]
+
+    # Create push server
+    if KEY_PUSH_SERVER not in hass.data[DOMAIN]:
+        push_server = PushServer(host)
+        hass.data[DOMAIN][KEY_PUSH_SERVER] = push_server
+        # start the async push server (only once)
+        await push_server.start()
+
+        # register stop callback to shutdown the push server
+        def stop_push_server(event):
+            """Stop push server."""
+            _LOGGER.debug("Shutting down Xiaomi Miio push server")
+            push_server.stop()
```

Comment:
Should this run in the event loop?

Reply:
I don't think so, it is doing non async IO to the registered devices in order to notify them to disconnect from the server.

However since the callback is not async, how can I schedule it in the executer thread?

Reply:
The asyncio transport close call should run in the event loop. The library needs to be refactored so that all library user interfaces are meant to be called from the event loop. If some internal calls in the library should run in another thread, the library needs to handle that.

Reply:
Alright, that seems very reasonable.
I will adjust the upstream library to make all push_server calls async.

Reply:
@MartinHjelmare Alright I converted all the upsteam push_server library functions to async.
That schould make it a bit more clear.

Reply:
Done